### PR TITLE
raftstore: optimize entry cache eviction to prevent append rejections under memory pressure

### LINKED ISSUE #17392: Optimize suspending writes when memory usage is high
## Background 
Here are 3 tikv

- tikv-2: Has the least remaining disk space (20G) < 5%, with some 'AlmostFull' errors, it refuse the write, and also other nodes do not send append log messages to this node (this issue lacks monitoring for handling proposals correctly if the majority of peers are disk full, as noted in Pull Request [handle proposals correctly if majority peers are disk full by hicqu · Pull Request #10671 · tikv/tik](https://github.com/tikv/tikv/pull/10671)).
image

- tikv0: because tikv-2 not push ahead raftlog, so tikv0, tikv-0 cannot compact log, the memory increase. after high memory usage in tikv0, it refuse to write(append):
image
image

- tikv1: also cannot server write because other 2 node stop append. 
 
image
image

##  What do you want to see
Improve this case: We need to prevent the expansion of this failure radius. the RAC is one node disk full, but the others are not full.  which lets another node's memory increase and another node's server because busy, it is unreasonable. if `tikv-0` can evict the entry cache or compaction log force, then `tikv-0` will not cause fault propagation, thereby reducing the explosion radius.

Improve the observability: the tidb side and client side see some error backoff is regionMiss or regionUnabliable, from the log, it report server is busy from `tikv-2`, the root cause is actually from the `tikv-1` with disk full.

## Version
v6.1.3

### PR DESCRIPTION
**Pull request by hhwyt**
### What is changed and how it works?

Issue Number: Close #17392, #17537

What's Changed:

```commit-message
This PR addresses the issue where, under memory pressure, a follower  
directly rejects append requests instead of first attempting to free  
up memory by evicting the entry cache.  

One potential solution is for the follower, before rejecting append  
requests, to notify other peers on the same node (including leaders  
and followers) to evict their entry caches. This would require a  
global module that coordinates peers. When many followers on the  
same node receive append requests under memory pressure, they would  
notify the module, which would then send messages to a sufficient  
number of peers to trigger entry cache eviction and report the  
results back to the followers. This module could also be triggered  
by a leader on the same node when it's receiving write requests  
under memory pressure. However, this solution is a bit complex and  
has a low ROI.  

A more practical solution is to optimize the conditions for  
triggering entry cache eviction, so it occurs earlier, avoiding  
append rejection.  

Why doesn't the current version trigger cache eviction earlier?  

Currently, entry cache eviction is checked either by the  
`raft_log_gc_tick_interval` (default 3 seconds) or during  
`handle_raft_committed_entries`. The trigger conditions for eviction  
are:  
1. Total memory usage has reached the high water mark.  
2. Entry cache memory usage has reached the  
`evict_cache_on_memory_ratio` (default 0.1, not visible to the  
customer).  

On the other hand, append rejection is triggered when:  
1. Total memory usage has reached the high water mark.  
2. The total memory usage of the entry cache, raft messages, and  
applying entries has reached the `reject_messages_on_memory_ratio`  
(default 0.2).  

The issue is that when the first condition is met, the second  
condition for append rejection may be triggered earlier.  

The solution proposed in this PR is to modify the first condition  
for cache eviction. Instead of waiting for memory usage to fully  
reach the high water mark, eviction will be triggered when memory  
usage is **near** the high water mark.  

This change should not introduce significant performance overhead  
because eviction is only triggered when memory usage is near the  
high water mark, and it helps prevent the more disruptive append  
rejections.  
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/entry_storage.rs [hunk 1]
```diff
@@ -1272,6 +1272,8 @@ impl<EK: KvEngine, ER: RaftEngine> EntryStorage<EK, ER> {
 
     /// Evict entries from the cache.
     pub fn evict_entry_cache(&mut self, half: bool) {
+        fail_point!("evict_entry_cache", |_| {});
```
[the remark this answers is not in the document]
Comment (hhwyt) [reply components/raftstore/src/store/entry_storage.rs]: done

File: components/raftstore/src/store/memory.rs [hunk 2]
```diff
@@ -57,18 +57,28 @@ lazy_static! {
         MEMTRACE_ROOT.sub_trace(Id::Name("raft_entries"));
 }
 
+pub fn get_memory_usage_entry_cache() -> u64 {
+    (|| {
+        fail_point!("memory_usage_entry_cache", |t| {
```
[the remark this answers is not in the document]
Comment (hhwyt) [reply components/raftstore/src/store/memory.rs]: done

**Review by LykxSassinator — APPROVED** [review]
Overall LGTM

**Review by glorv — APPROVED** [review]

**Review by hicqu — APPROVED** [review]

**Pull request merged**